### ISSUE TITLE
Some confuse about the decoder

### ISSUE THREAD

**Issue by tangtaogo** [YifYang993/CR-NeRF-PyTorch#2]
Thank you for your great work and open source code. However, I am confused about the code:
1. local variable 'net' referenced before assignment in https://github.com/YifYang993/CR-NeRF-PyTorch/blob/334fa6b591eb03d2be666b7ae7741fc0d206b178/models/nerf_decoder_stylenerf.py#L438
2. the parameter "decoder_num_res_blocks" is not being used in: https://github.com/YifYang993/CR-NeRF-PyTorch/blob/334fa6b591eb03d2be666b7ae7741fc0d206b178/models/linearStyleTransfer.py#L279
3. in the article, it is mentioned to set "m" to be smaller than "hcr1hcr2". However, in the code, it appears that the decoder receives image features that are of the same size as the image: https://github.com/YifYang993/CR-NeRF-PyTorch/blob/334fa6b591eb03d2be666b7ae7741fc0d206b178/models/linearStyleTransfer.py#L283
4. moreover, may I ask if I am correct in understanding that when using a decoder, the loss can only be calculated for the entire image, since the decoder generates the complete image every time, and cannot sample rays like traditional NERF to calculate the loss on discrete rays?

**Comment by YifYang993** [comment]
1,2: There are some classes that we defined but do not use in our CR-NeRF. I recommend using a debug mode to see which class and function are used. In our final version, our decoder uses the class in 
https://github.com/YifYang993/CR-NeRF-PyTorch/blob/334fa6b591eb03d2be666b7ae7741fc0d206b178/models/nerf_decoder_stylenerf.py#L227
3. Dimensionality and structure of our decoder: Please refer to images/architecture/decoder.pdf for a detailed feature dimension and architecture of the decoder. The parameters "img_size" and "featmap_size" only affect how many blocks we use in our decoder. Refer to 
https://github.com/YifYang993/CR-NeRF-PyTorch/blob/334fa6b591eb03d2be666b7ae7741fc0d206b178/models/nerf_decoder_stylenerf.py#L239
4. During training, we grid sample an image to a 32*32 image patch, and use rays that intersect with the patch for training, so the loss can be used for the patch and can also be used for the whole image. Note that during inference, we use the decoder to output the whole image simultaneously.

**Comment by tangtaogo** [comment]
Thank you for your prompt reply and clarification. 
1.2: Regarding the first question, I apologize for providing the wrong link in my previous response. Based on the given train.sh file, if we have --encode_a and --encode_c, it will lead to the code snippet at https://github.com/YifYang993/CR-NeRF-PyTorch/blob/334fa6b591eb03d2be666b7ae7741fc0d206b178/train_mask_grid_sample.py#L95. However, it seems that the parameter "decoder_num_res_blocks" is not being used in this function. Moving on to the decoder part, it is defined in https://github.com/YifYang993/CR-NeRF-PyTorch/blob/334fa6b591eb03d2be666b7ae7741fc0d206b178/models/nerf_decoder_stylenerf.py#L227. Within this decoder, there is an issue where the local variable 'net' is referenced before assignment in https://github.com/YifYang993/CR-NeRF-PyTorch/blob/334fa6b591eb03d2be666b7ae7741fc0d206b178/models/nerf_decoder_stylenerf.py#L282
3.4: Thank you for providing the excellent visualization, it is very clear. 
But here, img_size=(args.img_wh[0],args.img_wh[1]) and featmap_size=(args.img_wh[0],args.img_wh[1]) in https://github.com/YifYang993/CR-NeRF-PyTorch/blob/334fa6b591eb03d2be666b7ae7741fc0d206b178/models/linearStyleTransfer.py#L283, so does self.n_blocks = int(log2(img_size[0]/featmap_size[0])) always equal to 1?
And I have a question regarding the input being a 3x224x224 image and the decoder output being a 3x32x32 image patch. How should I calculate the RGB loss in this case?

**Comment by YifYang993** [comment]
1, 2. Since self.n_blocks equal to 0,  only this line executes

https://github.com/YifYang993/CR-NeRF-PyTorch/blob/334fa6b591eb03d2be666b7ae7741fc0d206b178/models/nerf_decoder_stylenerf.py#L280

you can ignore line 281 to line 287.
3, 4. Yes, self.n_blocks always equal to 0, you can try a deeper decoder to find if it can boost performance. 
Please try to debug the code, the input and output are both 32x32 during training. 
To be specific, suppose the input image is 3x224x224, after grid sampling, only a patch $P$ of 3x32x32  is selected. The camera rays are also selected based on the uv coordinate of the $P$, so the dimensions of the rays are nx32x32, where n is the number of ray points of a single ray. After feeding the rays into our CR-NeRF, we obtain an estimated 3x32x32 image patch $\hat{P}$, which is supervised by $P$. The dimensions of  $P$ and $\hat{P}$ are the same, so we can use L1 or L2 loss to minimize the discrepancy.

**Comment by YifYang993** [comment]
About how to choose m rays, please refer to our code datasets/phototourism_mask_grid_sample.py line 243 to line 266. We uniformly sample the m rays, and the m rays match the m corresponding pixel colors.

**Comment by tangtaogo** [comment]
Thank you for your patient reply and clarification !
